File: src/cpp/puzzlebuilder.cpp
#include "puzzlebuilder.h"

PuzzleBuilder::PuzzleBuilder()
	:m_permutationFactory(),
	 m_slotFactory(),
	 m_stickerFactory(),
	 m_puzzleData(NULL),
	 m_nodeMap()
{
}

PuzzleBuilder::~PuzzleBuilder()
{
}

void PuzzleBuilder::beginPuzzle(const string &name)
{
	m_puzzleData = new PuzzleData(name);
}

PuzzleData *PuzzleBuilder::compilePuzzle()
{
	return m_puzzleData;
}

void PuzzleBuilder::addNode(const string &name, const Color &color, const vec2f &center)
{
	Sticker *newSticker = m_stickerFactory.makeSticker(color);
	Slot *newSlot = m_slotFactory.makeSlot(color, center);
	newSticker->moveToSlot(newSlot);

	m_puzzleData->addSlot(newSlot);
	m_puzzleData->addSticker(newSticker);

	if (m_nodeMap.find(name) == m_nodeMap.end())
		m_nodeMap[name] = m_nodeMap.size();
}

void PuzzleBuilder::addPermutation(const Color &color, const vector<vector<string>> &cycleList)
{
	vector<vector<unsigned int>> indexedCycleList;
	
	for (const vector<string> &cycle : cycleList)
	{
		vector<unsigned int> indexedCycle;
		
		for (const string &name : cycle)
		{
			unsigned int index = m_nodeMap[name];
			indexedCycle.push_back(index);
		}
		
		indexedCycleList.push_back(indexedCycle);
	}
	
	m_permutationFactory.makePermutation(m_nodeMap.size(),
	                                     color,
	                                     0,
	                                     indexedCycleList);

	// TODO: PuzzleBuilder::addPermutation() : make a default line arc
}

void PuzzleBuilder::makeCircleArc(const Color &,
                                  const string &,
                                  const string &,
                                  const vec2f &)
{
	// TODO: PuzzleBuilder::makeCircleArc()
}

void PuzzleBuilder::makeLineArc(const Color &,
                                const string &,
                                const string &)
{
	// TODO: PuzzleBuilder::makeLineArc()
}

File: src/js/puzzlebuilder.js
class PuzzleBuilder
{
	constructor(name)
	{
		this.name = name;
		this.puzzleData = new PuzzleData(name);
		this.nodeList = [];
		this.permutationMap = {};
	}

	getPuzzleData()
	{
		this.reverseIndexList();
		return this.puzzleData;
	}

	reverseIndexList()
	{
		for (var i = 0; i < this.puzzleData.getPermutationListSize(); i++)
		{
			var j = this.puzzleData.getPermutationListSize() - i - 1;
			this.puzzleData.getPermutation(i).setReverseIndex(j);
		}
	}

	indexOf(nodeName)
	{
		for (var i = 0; i < this.puzzleData.getSlotListSize(); i++)
			if (this.puzzleData.getSlot(i).name == nodeName)
				return i;
		
		return -1;
	}

	addNode(nodeName, nodeColor, nodeShape, nodeX, nodeY)
	{
		var newSticker = new Sticker(nodeColor, nodeShape % NUM_SHAPES);
		var newSlot = new Slot(nodeName, newSticker, new vec2f(nodeX, nodeY), nodeColor);
		newSticker.moveToSlot(newSlot);
		this.puzzleData.addSlot(newSlot);
		this.puzzleData.addSticker(newSticker);
		this.nodeList.push(nodeName);
	}

	addPermutation(color, cycleList)
	{
		var n = this.puzzleData.slotList.length;
		var index = this.puzzleData.getPermutationListSize();
		var newPermutation = new Permutation(n, color, index);
		newPermutation.setCycles(cycleList);
		this.puzzleData.addPermutation(newPermutation);
		this.permutationMap[color] = newPermutation;
	}

	addCircleArc(arcColor, nodeName, circleR,
				 circlePlus, circleInverted, circleReversed, circleReflected)
	{
		var permutation = this.permutationMap[arcColor];
		var indexU = this.indexOf(nodeName);
		var indexV = permutation.next(indexU);
		var slotU = this.puzzleData.getSlot(indexU);
		var slotV = this.puzzleData.getSlot(indexV);
		var circleArc = NewCircleArc(permutation, slotU, slotV,
									 circleR, circlePlus, circleInverted,
									 circleReversed, circleReflected);
		this.puzzleData.setArc(permutation, indexU, circleArc);
	}
	
	recenter()
	{
		var centerSum = new vec2f(0.0, 0.0);

		for (var i = 0; i < this.puzzleData.getSlotListSize(); i++)
		{
			var slot = this.puzzleData.getSlotList()[i];
			
			var slotCenter = slot.getCenter();
			centerSum = centerSum.add(slotCenter);
		}

		centerSum = centerSum.scale(1.0 / this.puzzleData.getSlotListSize());

		for (var i = 0; i < this.puzzleData.getSlotListSize(); i++)
		{
			var slot = this.puzzleData.getSlotList()[i];

			var slotCenter = slot.getCenter();
			slotCenter = slotCenter.sub(centerSum);
			slot.setCenter(slotCenter);
		}

		for (var i = 0; i < this.puzzleData.getPermutationListSize(); i++)
		{
			var permutation = this.puzzleData.getPermutationList()[i];
			var arcList = this.puzzleData.getArcList(permutation);
			for (var j = 0; j < arcList.length; j++)
			{
				var arc = arcList[j];
				arc.adjustCenter(centerSum);
			}
		}
	}

	normalize()
	{
		var maxRadius = -1.0;

		for (var i = 0; i < this.puzzleData.getSlotListSize(); i++)
		{
			var slot = this.puzzleData.getSlotList()[i];

			var slotCenter = slot.getCenter();
			var radius = slotCenter.norm();
			if (radius > maxRadius)
				maxRadius = radius;
		}

		for (var i = 0; i < this.puzzleData.getSlotListSize(); i++)
		{
			var slot = this.puzzleData.getSlotList()[i];

			var slotCenter = slot.getCenter();
			slotCenter = slotCenter.scale(1.0 / maxRadius);
			slot.setCenter(slotCenter);
		}

		for (var i = 0; i < this.puzzleData.getPermutationListSize(); i++)
		{
			var permutation = this.puzzleData.getPermutationList()[i];
			var arcList = this.puzzleData.getArcList(permutation);
			for (var j = 0; j < arcList.length; j++)
			{
				var arc = arcList[j];
				arc.normalize(maxRadius);
				arc.computeParameters();
			}
		}
	}
}
